componentLibraries/defaultLibrary/Special/AeroComponents/AeroFlightRecorder.hpp

```diff
@@ -0,0 +1,329 @@
+                         "  <open>1</open>\n"
+                         "  <Style id=\"redLineRedPoly\">\n"
+                         "    <PolyStyle>\n"
+                         "      <color>4C0000FF</color>\n"
+                         "      <colorMode>normal</colorMode>\n"
+                         "    </PolyStyle>\n"
+                         "    <LineStyle>\n"
+                         "      <color>4C0000FF</color>\n"
+                         "      <colorMode>normal</colorMode>\n"
+                         "      <width>4</width>\n"
+                         "    </LineStyle>\n"
+                         "  </Style>\n"
+                         "  <Folder>\n"
+                         "    <name>SimResults</name>\n"
+                         "    <Placemark>\n"
+                         "      <name>Ground Path</name>\n"
+                         "      <styleUrl>#redLineRedPoly</styleUrl>\n"
+                         "      <LineString>\n"
+                         "        <altitudeMode>clampToGround</altitudeMode>\n"
+                         "        <coordinates>\n";
+            }
+
+            else if(mFileType == ThreeDmodelKML) {
+                mFile.open((mFilePath.append("\\"+modelName+".kml")).c_str());
+                if (!mFile.is_open()) {
+                    stopSimulation("Could not find, open, or write an output file in the given path: "+ mFilePath);
+                    return;
+                }
+                mFile << "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+                         "<kml xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\">\n"
+                         "<Document>\n"
+                         "  <open>1</open>\n"
+                         "  <Folder>\n"
+                         "    <name>SimResults</name>\n"
+                         "    <Folder>\n"
+                         "      <name>3-D Aircraft</name>\n";
+            }
+
+            // Call per-step function
+            mLastLogTime = mTime-(*mpDt);
+            simulateOneTimestep();
+        }
+
+
+
+        void simulateOneTimestep()
+        {
+            //Check if it is time to run
+            if(mTime < mLastLogTime+(*mpDt)) {
+                return;
+            }
+            mLastLogTime = mTime;
+
+            //Read variables from nodes
+            longitude = (*mpLongitude);
+            latitude = (*mpLatitude);
+            altitude = (*mpAltitude);
+
+            //Write the data according to selected type
+            if(mFileType == FlightPathKML) {
+                mFile << "        " << longitude << "," << latitude << "," << altitude << "\n";
+            }
+
+            else if(mFileType == GroundPathKML) {
+                mFile << "        " << longitude << "," << latitude << "," << altitude << "\n";
+            }
+
+            else if(mFileType == ThreeDmodelKML) {
+                roll = (*mpPhi) * -57.29577951;
```

Comment:
What are these magic values ?
57.29577951
3.6

Reply:
I guess 57.29... is 180/pi. It could perhaps be replaced with the `deg2rad` function from `ComponentUtilities.h` to make the code more readable.

Reply:
@robbr48 is right, it's just a manual `rad2deg` as I wasn't aware that such a function was available. I'll change it as it is much more readable with the function.

The 3.6 is a "m/s" to "km/h". Not sure if there is a function for that, but at least I'll comment it in the code so it's clear what is going on. Or even better, I just leave it with the same units as it comes. It's not essential to get "km/h", as these values are just presented as information in the KML (possibly, in the future, they will play a role in a flight "replay" animation).

Reply:
Just leave a comment in the code. 3.6 = m/s to km/h, or similar